### ISSUE TITLE
Performance when using hundreds of relationships

### ISSUE THREAD

**Issue by benbosman** [DSpace/DSpace#3125]
**Describe the bug**
We've performed a performance test creating a Person entity with hundreds of relationships to various Publications. This is a realistic use-case, e.g. if the Person is a high-profile researcher who's co-author of many publications in their department.

We noticed the duration for adding e.g. the one-hundred or three-hundred relationship is much higher compared to the first relationship

**To Reproduce**
1. Start with a database with many publications
2. Iterate the publications, and add a relationship to the same Person
3. Verify the response time when adding a relationship

**Expected behavior**
The performance can be optimized a lot by excluding the generation of virtual metadata on some relationships where it's not relevant. The virtual metadata in the Person item containing all publications is currently not used, excluding it can help a lot

**Related work**
https://github.com/DSpace/DSpace/pull/2895 has initially been started for this improvement, but hit some issues. Although the concept is still good, the implementation needs to be revised

**Comment by tdonohue** [comment]
@benbosman : Thanks for the ticket here. This sounds like it would need more analysis/discussion (which may need to wait for 7.1), unless you are suggesting someone (quickly) update #2895 (as a quick fix) in time for Beta 5?

**Comment by benbosman** [comment]
@tdonohue there's WIP on this ticket, @cwilper is planning to create a new PR for this work
We're also identifying other improvements which can easily be completed in time for beta 5

**Comment by tdonohue** [comment]
@benbosman : OK, I'll tentatively add this to Beta 5.  However, I'd like a clearer estimate / scope for this ticket.  I'm (obviously) not wanting to delay Beta 5 just for this work -- that said, as you know Beta 5 still is waiting on theming work.

**Comment by benbosman** [comment]
I will work with @cwilper to add some further detail on this ticket.
I understand there's no plan to delay beta 5 for this work, but it can be worked on in the meanwhile (we'll be talking different people who're involved anyway)

**Comment by benbosman** [comment]
This ticket is split in 3 PRs:
* https://github.com/DSpace/DSpace/pull/3131 is a bugfix which required 3 hours of analysis and 1 hour of implementation
* https://github.com/DSpace/DSpace/pull/3134 drastically improves performance for relationships which don't need to be bi-directional. It required 15 hours of work (there's been many iterations of the implementation), and is estimated to require 4 hours to finalize
* https://github.com/DSpace/DSpace/pull/3170 further improves performance substantially. It required 3 hours of analysis and 4 hours of implementation

The combination of these changes improves performance drastically, making the system very response even with large amounts of relationships

**Comment by tdonohue** [comment]
Fixed by #3134 and #3170
